exporter/cortexexporter/DESIGN.md

```diff
@@ -0,0 +1,287 @@
+          export(map)  
+       }
+
+### **2.2 Mapping of OTLP Metrics to TimeSeries**
+
+Each Cortex TimeSeries represents less semantic information than an OTLP metric. The temporality property of a OTLP metric is ignored in a TimeSeries because it is always considered as cumulative for monotonic types and histogram, and the type property of a OTLP metric is translated by mapping each metric to one or multiple TimeSeries. The following sections explain how to map each OTLP metric type to Cortex TimeSeries.
+
+
+**INT64, MONOTONIC_INT64, DOUBLE, MONOTONIC_DOUBLE**
+
+Each unique label set within metrics of these types can be converted to exactly one TimeSeries. From the perspective of Prometheus Client types, INT64 and DOUBLE correspond to gauge metrics, and MONOTONIC types correspond to counter metrics. In both cases, data points will be exported directly without aggregation. Any metric of the monotonic types that is not cumulative should be dropped; non-monotonic scalar types are assumed to represent gauge values, thus its temporality is not checked. Monotonic types need to have a `_total` suffix in its metric name when exporting; this is a requirement of [Prometheus](https://www.slideshare.net/brianbrazil/openmetrics-what-does-it-mean-for-you-promcon-2019-munich) (slide 11).
+
+
+**HISTOGRAM**
+
+Each histogram data point can be converted to 2 + n + 1 Cortex TimeSeries:
+
+* 1 *TimeSeries* representing metric_name_count contains HistogramDataPoint.count
+* 1 *TimeSeries* representing metric_name_sum contains HistogramDataPoint.sum
+* n *TimeSeries* each representing metric_name_bucket{le=“upperbound”} contain the count of each bucket defined by the bounds of the data point
+* 1 *TimeSeries* representing metric_name_bucket{le=“+Inf”} contains counts for the bucket with infinity as upper bound; its value is equivalent to metric_name_count.
+
+Cortex bucket values are cumulative, meaning the count of each bucket should contain counts from buckets with lower bounds. In addition, Exemplars from a histogram data point are ignored. When adding a bucket of the histogram data point to the map, the string signature should also contain a `le` label that indicates the bound value. This label should also be exported. Any histogram metric that is not cumulative should be dropped.
+
+
+**SUMMARY**
+
+Each summary data point can be converted to 2 + n Cortex TimeSeries:
+
+* 1 *TimeSeries* representing metric_name_count contains SummaryDataPoint.count
+* 1 *TimeSeries* representing metric_name_sum contains SummaryDataPoint.sum
+* and n *TimeSeries* each representing metric_name{quantile=“quantileValue”} contains the value of each quantile in the data point.
+
+When adding a quantile of the summary data point to the map, the string signature should also contain a `quantile ` label that indicates the quantile value. This label should also be exported. Any summary metric that is not cumulative should be dropped.
+
+### **2.3 Exporting Metrics**

```

Comment:
Would it be easier to vendor in Prometheus remote write package here?

Reply:
Thanks for the suggestion, using the client from remote write makes sense. We are also trying to allow users to specify any header or pass in a http.Client to send requests. I will take a look at whether this is possible with remote write package.

Reply:
It doesn't sadly, but we should see if we can make the current client accept a custom `http.Client`. 

https://github.com/prometheus/prometheus/blob/master/storage/remote/client.go#L125-L133

We can try adding a `SetHTTPClient()` method to the Prometheus client object.

Reply:
Yes, I think that change would be very helpful for our usage.